lib/Re2.fx

```diff
@@ -0,0 +1,828 @@
+        *fx_result = re2::RE2::FullMatch(ctext.data, *re_to_apply);//TODO: Try - catch.
+        fx_free_cstr(&ctext);
+    }
+    return fx_status;
+}
+
+//                                                 (success, (sub_start, sub_end)[])
+pure fun full_match_n(text: string, re: regex_t) : (bool   , (int      , int    )[]) = ccode
+{   
+    using namespace re2;
+    fx_cstr_t ctext;
+    int fx_status = fx_str2cstr(text, &ctext, 0, 0);
+    if(fx_status>=0)
+    {
+        RE2* re_to_apply = static_cast<RE2*>(re->handle->ptr);
+        const int sub_amount = re_to_apply->NumberOfCapturingGroups();
+
+        StringPiece* sub_string_pieces;
+        RE2::Arg** sub_args;
+        int fx_status = make_sub_args(sub_string_pieces, sub_args, sub_amount);
+        if(fx_status>=0)
+        {
+            fx_result->t0 = RE2::FullMatchN(ctext.data, *re_to_apply, sub_args, sub_amount); //TODO: try - catch!
+            if(fx_result->t0)
+            {
+                fx_status = string_pieces_to_arr(sub_string_pieces, sub_amount, ctext.data, &(fx_result->t1));
+            }
+            free_sub_args(sub_string_pieces, sub_args, sub_amount);
+        }
+        fx_free_cstr(&ctext);
+    }
+    return fx_status;
+}
+
+//                                                     (success, sub-matches [])
+pure fun full_match_n_str(text: string, re: regex_t) : (bool,         string [])
+{   
+    val (success, starts_ends) = full_match_n(text, re)
+    (success, incise(text, starts_ends))
+}
+
+pure fun partial_match(text: string, re: regex_t) : bool = ccode
+{
+    using namespace re2;
+    fx_cstr_t ctext;
+    int fx_status = fx_str2cstr(text, &ctext, 0, 0);
+    if(fx_status>=0)
+    {
+        RE2* re_to_apply = static_cast<RE2*>(re->handle->ptr);
+        *fx_result = re2::RE2::PartialMatch(ctext.data, *re_to_apply);//TODO: Try - catch.
+        fx_free_cstr(&ctext);
+    }
+    return fx_status;
+}
+
+//                                                    (success, (sub_start, sub_end)[])
+pure fun partial_match_n(text: string, re: regex_t) : (bool   , (int      , int    )[]) = ccode
+{   
+    using namespace re2;
+    fx_cstr_t ctext;
+    int fx_status = fx_str2cstr(text, &ctext, 0, 0);
+    if(fx_status>=0)
+    {
+        RE2* re_to_apply = static_cast<RE2*>(re->handle->ptr);
+        const int sub_amount = re_to_apply->NumberOfCapturingGroups();
+
+        StringPiece* sub_string_pieces;
+        RE2::Arg** sub_args;
+        int fx_status = make_sub_args(sub_string_pieces, sub_args, sub_amount);
+        if(fx_status>=0)
+        {
+            fx_result->t0 = RE2::PartialMatchN(ctext.data, *re_to_apply, sub_args, sub_amount); //TODO: try - catch!
+            if(fx_result->t0)
+            {
+                fx_status = string_pieces_to_arr(sub_string_pieces, sub_amount, ctext.data, &(fx_result->t1));
+            }
+            free_sub_args(sub_string_pieces, sub_args, sub_amount);
```

Comment:
since the body is C++ function I'd suggest to use C++ containers with automatic destructors more extensively instead of deeply-nested if's. Of course, RE2 exceptions, if any, need to be caught and converted to ficus exceptions.

Reply:
Done.